File: KingPastaKit/app/period.rb
class Period
  include MotionModel::Model
  include MotionModel::ArrayModelAdapter
  include MotionModel::Validatable

  columns :name => :string,
          :startTime => :date,
          :endTime => :date

  # validates :name, :presence => true
  # validates :startTime, :presence => true
  # validates :endTime, :presence => true

  belongs_to :schedule
  belongs_to :category

  def self.current
    time = Time.now.strip_date

    if periods_today = Period.all_today then periods_today.where(:startTime).lte(time).where(:endTime).gt(time).first end
  end

  def self.next
    time = Time.now.strip_date

    if periods_today = Period.all_today then periods_today.where(:startTime).gt(time).first end
  end

  def self.all_today
    if schedule_today = Schedule.today then schedule_today.all_periods end
  end

  def self.all_on_wday(wday)
    if schedule_wday = Schedule.on_wday(wday) then schedule_wday.all_periods end
  end

  def self.overlap_with_start(startTime, end: endTime, schedule: schedule, ignoring: old_period)
    start_during = schedule.periods.where(:startTime).gte(startTime).where(:startTime).lt(endTime).array.reject { |p| p == old_period }
    end_during = schedule.periods.where(:endTime).gt(startTime).where(:endTime).lte(endTime).array.reject { |p| p == old_period }
    before_to_after = schedule.periods.where(:startTime).lte(startTime).where(:endTime).gte(endTime).array.reject { |p| p == old_period }
    
    (start_during.count + end_during.count + before_to_after.count) > 0
  end

  def time_remaining
    Time.at(self.endTime - Time.now.strip_seconds).utc
  end

  def time_until_start
    Time.at(self.startTime - Time.now.strip_seconds).utc
  end

end

File: KingPastaKit/Period.swift
import Foundation
import RealmSwift

public class Period: Object {
    public dynamic var name: String = ""
    public dynamic var startTime: NSTimeInterval = 0
    public dynamic var endTime: NSTimeInterval = 0
    
    public dynamic var day: Day?
    public dynamic var category: KPCategory?
    
    // TODO: This can be achieved better
    convenience init(name: String, startTime: NSTimeInterval, endTime: NSTimeInterval, day: Day, category: KPCategory) {
        self.init()
        
        self.name = name
        self.startTime = startTime
        self.endTime = endTime
        self.day = day
        self.category = category
    }
    
    class func current() -> Period? {
        guard let periodsToday = Period.allToday() else { return nil }
        let time = Time.now()
        return periodsToday.filter("startTime <= %@", time).filter("endTime > %@", time).first
    }
    
    class func next() -> Period? {
        guard let periodsToday = Period.allToday() else { return nil }
        let time = Time.now()
        return periodsToday.filter("startTime > %@", time).first
    }
    
    class func allToday() -> Results<Period>? {
        guard let today = Day.today() else { return nil }
        return today.orderedPeriods()
    }
    
    class func allOnDayOfWeek(dayOfWeek: Int) -> Results<Period>? {
        guard let day = Day.getDayOfWeek(dayOfWeek) else { return nil }
        return day.orderedPeriods()
    }
    
    public class func overlapWithStartTime(startTime: NSTimeInterval, endTime: NSTimeInterval, day: Day?, ignoringPeriod: Period?) -> Bool {
        guard let day = day else { return true }
        
        // TODO: is the last filter with block acceptable for Results<T>???
        let startDuring = day.periods.filter("startTime >= %@", startTime).filter("startTime < %@", endTime).filter { $0 != ignoringPeriod }
        let endDuring = day.periods.filter("endTime > %@", startTime).filter("endTime <= %@", endTime).filter { $0 != ignoringPeriod }
        let beforeToAfter = day.periods.filter("startTime <= %@", startTime).filter("endTime >= %@", endTime).filter { $0 != ignoringPeriod }
        
        return (startDuring.count + endDuring.count + beforeToAfter.count) > 0
    }
    
    public enum OverlapType {
        case StartsAtEnd
        case StartsAfterEnd
        case Overlap
    }

    var timeRemaining: NSTimeInterval {
        return endTime - Time.now().approx
    }
    
    var timeUntilStart: NSTimeInterval {
        return startTime - Time.now().approx
    }
}